app/models/bundle_context.rb

```diff
@@ -98,21 +105,48 @@ def manifest_rows
 
   def manifest_cols
     {
-      label: 'label',
-      source_id: 'sourceid',
+      label: 'label', # only used by SMPL manifests
+      source_id: 'sourceid', # only used by SMPL manifests
       object_container: 'object', # object referring to filename or foldername
       druid: 'druid'
     }
   end
 
   private
 
+  def job_output_parent_dir
+    @job_output_parent_dir ||= Settings.job_output_parent_dir
+  end
+
+  def normalize_dir(dir)
+    dir.chomp('/') if dir
+  end
+
+  def normalize_bundle_dir
+    self[:bundle_dir] = normalize_dir(bundle_dir)
+  end
+
+  def verify_output_dir
+    if persisted?
+      # FIXME: or should we just try to create it?
+      raise "Output directory (#{output_dir}) should already exist, but doesn't" unless Dir.exist?(output_dir)
+    else
```

Comment:
I think you might get more clarity by using a `before_create` callback to target specifically the not-`persisted?` condition.  Then the `before_create` callback for does the actual creation and the `before_save` still ensures existence and testing the respective methods is a simpler affair.

Reply:
Hmm, I guess that is complicated by the fact `before_save` would fire *before* `before_create` on `BundleContext.new(...).save`:
https://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html

So `before_save` would still have to consider `persisted?`, but with a distinct method, that check could be attached to the callback declaration, like `before_save :verify_output_dir, if: -> { persisted? }`.  Does that make sense?

Reply:
#316